k-distribution/k-tutorial/1_basic/03_parsing/README.md

```diff
@@ -1 +1,627 @@
+attributes together into a left-, right-, or non-associative grouping.
+
+## Prefer/avoid
+
+Sometimes priority and associativity proves insufficient to disambiguate a
+grammar. In particular, sometimes it is desirable to be able to choose between
+two ambiguous parses directly while still not rejecting any parses if the term
+parsed is unambiguous. A good example of this is the famous "dangling else"
+problem in imperative C-like languages.
+
+Consider the following definition (`lesson-03-i.k`):
+
+```k
+module LESSON-03-I
+
+  syntax Exp ::= "true" | "false"
+  syntax Stmt ::= "if" "(" Exp ")" Stmt
+                | "if" "(" Exp ")" Stmt "else" Stmt
+                | ";"
+endmodule
+```
+
+We can write the following program (`dangling-else.if`):
+
+```
+if (true) if (false) ; else ;
+```
+
+This is ambiguous because it is unclear whether the `else` clause is part of
+the outer `if` or the inner `if`. At first we might try to resolve this with
+priorities, saying that the `if` without an `else` cannot appear as a child of
+the `if` with an `else`. However, because the non-terminal in the parent symbol
+is both preceded and followed by a terminal, this will not work.
+
+Instead, we can resolve the ambiguity directly by telling the parser to
+"prefer" or "avoid" certain productions when ambiguities arise. For example,
+when we parse this program, we see the following ambiguity as an error message:
+
+```
+[Error] Inner Parser: Parsing ambiguity.
+1: syntax Stmt ::= "if" "(" Exp ")" Stmt
+
+`if(_)__LESSON-03-I_Stmt_Exp_Stmt`(`true_LESSON-03-I_Exp`(.KList),`if(_)_else__LESSON-03-I_Stmt_Exp_Stmt_Stmt`(`false_LESSON-03-I_Exp`(.KList),`;_LESSON-03-I_Stmt`(.KList),`;_LESSON-03-I_Stmt`(.KList)))
+2: syntax Stmt ::= "if" "(" Exp ")" Stmt "else" Stmt
+
+`if(_)_else__LESSON-03-I_Stmt_Exp_Stmt_Stmt`(`true_LESSON-03-I_Exp`(.KList),`if(_)__LESSON-03-I_Stmt_Exp_Stmt`(`false_LESSON-03-I_Exp`(.KList),`;_LESSON-03-I_Stmt`(.KList)),`;_LESSON-03-I_Stmt`(.KList))
+        Source(./dangling-else.if)
+        Location(1,1,1,30)
+```
+
+Roughly, we see that the ambiguity is between an `if` with an `else` or an `if`
+without an `else`. Siknce we want to pick the first parse, we can tell K to
+"avoid" the second parse with the `avoid` attribute. Consider the following
+modified definition (`lesson-03-j.k`):
+
+```k
+module LESSON-03-I
+
+  syntax Exp ::= "true" | "false"
+  syntax Stmt ::= "if" "(" Exp ")" Stmt
+                | "if" "(" Exp ")" Stmt "else" Stmt [avoid]
+                | ";"
+endmodule
+```
+
+Here we have added the `avoid` attribute to the `else` production. As a result,
+when an ambiguity occurs and one or more of the possible parses has that symbol
+at the top of the ambiguous part of the parse, we remove those parses from
+consideration and consider only those remaining. The `prefer` attribute behaves
+similarly, but instead removes all parses which do not have that attribute.
+In both cases, no action is taken if the parse is not ambiguous.
```

Comment:
```suggestion
Here we have added the `avoid` attribute to the `else` production. Now,
when an ambiguity occurs, any of the possible parses which has the `avoid` symbol
at the top will be removed from consideration. The `prefer` attribute behaves
similarly, but instead removes all parses which do not have a `prefer` symbol at the top.
In both cases, no action is taken if the parse is not ambiguous.
```

Reply:
It is actually a clearer explanation the way I originally left it. Your wording doesn't adequately explain that what matters is where the ambiguous part of the parse begins.